docs/architecture/adr-108-e2e-abci++.md

```diff
@@ -0,0 +1,305 @@
+that returns `abci.Request` from the string are provided in files `test/e2e/app/log.go` and `test/e2e/app/log_test.go`, respectively. To create a string representation of a request, we first marshal the request via `proto.Marshal()` method and then convert received bytes in the string using `base64.StdEncoding.EncodeToString()` method. In addition, we surround the new string with `abci-req` constants so that we can find lines with ABCI++ request more easily. The code of the method is below: 
+
+```go
+func GetABCIRequestString(req *abci.Request) (string, error) {
+	b, err := proto.Marshal(req)
+	if err != nil {
+		return "", err
+	}
+	reqStr := base64.StdEncoding.EncodeToString(b)
+	s := ABCI_REQ + reqStr + ABCI_REQ
+	return s, nil
+}
+```
+
+*Note:* At the moment, we are not compressing the marshalled request before converting it to `base64` `string` because we are logging the empty requests that take at most 24 bytes. However, if we decide to log the actual requests in the future, we might want to compress them. Based on a few tests, we observed that the size of a request can go up to 7KB.  
+
+If in the future we want to log another ABCI++ request type, we just need to do the same thing: 
+create a corresponding `abci.Request` and log it via 
+`app.logABCIRequest(r)`. 
+
+### 2) Parsing the logs
+We need a code that will take the logs from all nodes and collect the ABCI++ requests that were logged by the application. 
+
+**Implementation**
+
+This logic is implemented inside the `fetchABCIRequests(t *testing.T, nodeName string)` function that resides in `test/e2e/tests/e2e_test.go` file. This function does three things:
+- Takes the output of a specific node in the testnet from the moment we launched the testnet until the function is called. The node name is passed as a function parameter. It uses the `docker-compose logs` and `grep nodeName` commands. 
+- Parses the logs line by line and extracts the  `abci.Request`, if one exists. The request is received by forwarding each line to the `app.GetABCIRequestFromString(req)` method.
+- Returns the array of slices where each slice contains the set of `abci.Request`s logged on that node. Every time a crash happens, a new array element (new slice `[]*abci.Request`) will be created. We know a crash has happened because we log "Application started" every time the application starts. Specifically, we added this log inside `NewApplication()` function in `test/e2e/app/app.go` file. In the end, `fetchABCIRequests()` will return just one slice if the node did not experience any crashes and $n+1$ slices if there were crashes, $n$ being the number of crashes. The benefit of logging the requests in the previously described way is that now we can use `[]*abci.Request` to store ABCI++ requests of any type.
```

Comment:
```suggestion
- Returns the array of slices where each slice contains the set of `abci.Request`s logged on that node. Every time a crash happens, a new array element (new slice `[]*abci.Request`) will be created. We know a crash has happened because we log "Application started" every time the application starts. Specifically, we added this log inside `NewApplication()` function in `test/e2e/app/app.go` file. In the end, `fetchABCIRequests()` will return just one slice if the node did not experience any crashes and $n+1$ slices if there were $n$ crashes. The benefit of logging the requests in the previously described way is that now we can use `[]*abci.Request` to store ABCI++ requests of any type.
```
it's just a suggestion